jerry-ext/handle-scope/handle-scope.c

```diff
@@ -0,0 +1,373 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <stdlib.h>
+#include "handle-scope-internal.h"
+
+/**
+ * Opens a new handle scope and attach it to current global scope as a child scope.
+ *
+ * @param result - [out value] opened scope.
+ * @return status code, jerryx_handle_scope_ok if success.
+ */
+jerryx_handle_scope_status
+jerryx_open_handle_scope (jerryx_handle_scope *result)
+{
+  *result = jerryx_handle_scope_alloc ();
+  return jerryx_handle_scope_ok;
+} /** jerryx_open_handle_scope */
+
+
+/**
+ * Release all jerry values attached to given scope
+ *
+ * @param scope - the scope of handles to be released.
+ */
+void
+jerryx_handle_scope_release_handles (jerryx_handle_scope scope)
+{
+  size_t handle_count = scope->handle_count;
+  if (scope->handle_count > JERRYX_HANDLE_PRELIST_SIZE)
+  {
+    jerryx_handle_t *a_handle = scope->handle_ptr;
+    while (a_handle != NULL)
+    {
+      jerry_release_value (a_handle->jval);
+      jerryx_handle_t *sibling = a_handle->sibling;
+      free (a_handle);
+      a_handle = sibling;
+    }
+    handle_count = JERRYX_HANDLE_PRELIST_SIZE;
+  }
+
+  for (size_t idx = 0; idx < handle_count; idx++)
+  {
+    jerry_release_value (scope->handle_prelist[idx]);
+  }
+  scope->handle_count = 0;
+} /** jerryx_handle_scope_release_handles */
+
+
+/**
+ * Close the scope and its child scopes and release all jerry values that
+ * resides in the scopes.
+ * Scopes must be closed in the reverse order from which they were created.
+ *
+ * @param scope - the scope closed.
+ * @return status code, jerryx_handle_scope_ok if success.
+ */
+jerryx_handle_scope_status
+jerryx_close_handle_scope (jerryx_handle_scope scope)
+{
+  /**
+   * Release all handles related to given scope and its child scopes
+   */
+  jerryx_handle_scope a_scope = scope;
+  do
+  {
+    jerryx_handle_scope_release_handles (a_scope);
+    jerryx_handle_scope child = jerryx_handle_scope_get_child (a_scope);
+    jerryx_handle_scope_free (a_scope);
+    a_scope = child;
+  }
+  while (a_scope != NULL);
+
+  return jerryx_handle_scope_ok;
+} /** jerryx_close_handle_scope */
+
+
+/**
+ * Opens a new handle scope from which one object can be promoted to the outer scope
+ * and attach it to current global scope as a child scope.
+ *
+ * @param result - [out value] opened escapable handle scope.
+ * @return status code, jerryx_handle_scope_ok if success.
+ */
+jerryx_handle_scope_status
+jerryx_open_escapable_handle_scope (jerryx_handle_scope *result)
+{
+  return jerryx_open_handle_scope (result);
+} /** jerryx_open_escapable_handle_scope */
+
+
+/**
```

Comment:
There are lots of places with double newlines, but I think JerryScript uses single newlines between functions.

Reply:
Yes, JerryScript uses single newlines pattern. I'd update the style.